src/modules/dotnet.rs

```diff
@@ -0,0 +1,308 @@
+use std::ops::Deref;
+use std::path::Path;
+use std::process::{Command, Stdio};
+use std::str;
+
+use super::{Context, Module};
+
+type JValue = serde_json::Value;
+
+const DOTNET_SYMBOL: &str = "•NET ";
+const GLOBAL_JSON_FILE: &str = "global.json";
+const PROJECT_JSON_FILE: &str = "project.json";
+
+/// A module which shows the latest (or pinned) version of the dotnet SDK
+///
+/// Will display if any of the following files are present in
+/// the current directory:
+/// global.json, project.json, *.sln, *.csproj, *.fsproj, *.xproj
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let dotnet_files = get_local_dotnet_files(context).ok()?;
+    if dotnet_files.is_empty() {
+        return None;
+    }
+
+    let mut module = context.new_module("dotnet");
+
+    // Internally, this module uses its own mechanism for version detection.
+    // Typically it is twice as fast as running `dotnet --version`.
+    let enable_heuristic = module.config_value_bool("heuristic").unwrap_or(true);
+    let version = if enable_heuristic {
+        let repo_root = context
+            .get_repo()
+            .ok()
+            .and_then(|r| r.root.as_ref().map(|buf| buf.as_path()));
+        estimate_dotnet_version(&dotnet_files, &context.current_dir, repo_root)?
+    } else {
+        get_version_from_cli()?
+    };
+
+    let module_style = module
+        .config_value_style("style")
+        .unwrap_or_else(|| Color::Blue.bold());
+
+    module.set_style(module_style);
+    module.new_segment("symbol", DOTNET_SYMBOL);
+    module.new_segment("version", &version);
+
+    Some(module)
+}
+
+fn estimate_dotnet_version<'a>(
+    files: &[DotNetFile<'a>],
+    current_dir: &Path,
+    repo_root: Option<&Path>,
+) -> Option<Version> {
+    let get_file_of_type = |t: FileType| files.iter().find(|f| f.file_type == t);
+
+    // Check for a global.json or a solution file first, but otherwise take the first file
+    let relevant_file = get_file_of_type(FileType::GlobalJson)
+        .or_else(|| get_file_of_type(FileType::SolutionFile))
+        .or_else(|| files.iter().next())?;
+
+    match relevant_file.file_type {
+        FileType::GlobalJson => {
+            get_pinned_sdk_version_from_file(relevant_file.path).or_else(get_latest_sdk_from_cli)
+        }
+        FileType::SolutionFile => {
+            // With this heuristic, we'll assume that a "global.json" won't
+            // be found in any directory above the solution file.
+            get_latest_sdk_from_cli()
+        }
+        _ => {
+            // If we see a dotnet project, we'll check a small number of neighboring
+            // directories to see if we can find a global.json. Otherwise, assume the
+            // latest SDK is in use.
+            try_find_nearby_global_json(current_dir, repo_root).or_else(get_latest_sdk_from_cli)
+        }
+    }
+}
+
+fn try_find_nearby_global_json(current_dir: &Path, repo_root: Option<&Path>) -> Option<Version> {
```

Comment:
Can we get a `///` doc comment for this function?

Reply:
Yep. Done.